### ISSUE TITLE
r10k.bat fails on windows when executed remotely (Windows OpenSSH)

### ISSUE THREAD

**Issue by lmayorga1980** [puppetlabs/r10k#455]
Start an SSH Session from Linux -> Windows(OpenSSH Server)
### SSH Session

```
Powershell -file c:/temp/puppet-install-code.ps1
ERROR    -> Command exited with non-zero exit code:
Command: git clone --mirror https://github.com/puppetlabs/puppetlabs-stdlib.git C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-puppetlabs-puppetlabs-stdlib.git
Stderr:
Too many arguments.
```

This usually works when running it locally.

**Comment by adrienthebo** [comment]
Could you provide the output of running r10k with --trace?

**Comment by lmayorga1980** [comment]
```
Connecting <MYHOST>:22
15:45:13 [<MYHOST> 3][ERROR] ERROR   -> Command exited with non-zero exit code:
15:45:13 [<MYHOST> 3][ERROR] Command: git clone --mirror https://github.com/puppetlabs/puppetlabs-stdlib.git C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-puppetlabs-puppetlabs-stdlib.git
15:45:13 [<MYHOST> 3][ERROR] Stderr:
15:45:13 [<MYHOST> 3][ERROR] Too many arguments.
15:45:13 [<MYHOST> 3][ERROR] 
15:45:13 [<MYHOST> 3][ERROR] usage: git clone [options] [--] <repo> [<dir>]
15:45:13 [<MYHOST> 3][ERROR] 
15:45:13 [<MYHOST> 3][ERROR]     -v, --verbose         be more verbose
15:45:13 [<MYHOST> 3][ERROR]     -q, --quiet           be more quiet
15:45:13 [<MYHOST> 3][ERROR]     --progress            force progress reporting
15:45:13 [<MYHOST> 3][ERROR]     -n, --no-checkout     don't create a checkout
15:45:13 [<MYHOST> 3][ERROR]     --bare                create a bare repository
15:45:13 [<MYHOST> 3][ERROR]     --mirror              create a mirror repository (implies bare)
15:45:13 [<MYHOST> 3][ERROR]     -l, --local           to clone from a local repository
15:45:13 [<MYHOST> 3][ERROR]     --no-hardlinks        don't use local hardlinks, always copy
15:45:13 [<MYHOST> 3][ERROR]     -s, --shared          setup as shared repository
15:45:13 [<MYHOST> 3][ERROR]     --recursive           initialize submodules in the clone
15:45:13 [<MYHOST> 3][ERROR]     --recurse-submodules  initialize submodules in the clone
15:45:13 [<MYHOST> 3][ERROR]     --template <template-directory>
15:45:13 [<MYHOST> 3][ERROR]                           directory from which templates will be used
15:45:13 [<MYHOST> 3][ERROR]     --reference <repo>    reference repository
15:45:13 [<MYHOST> 3][ERROR]     -o, --origin <name>   use <name> instead of 'origin' to track upstream
15:45:13 [<MYHOST> 3][ERROR]     -b, --branch <branch>
15:45:13 [<MYHOST> 3][ERROR]                           checkout <branch> instead of the remote's HEAD
15:45:13 [<MYHOST> 3][ERROR]     -u, --upload-pack <path>
15:45:13 [<MYHOST> 3][ERROR]                           path to git-upload-pack on the remote
15:45:13 [<MYHOST> 3][ERROR]     --depth <depth>       create a shallow clone of that depth
15:45:13 [<MYHOST> 3][ERROR]     --single-branch       clone only one branch, HEAD or --branch
15:45:13 [<MYHOST> 3][ERROR]     --separate-git-dir <gitdir>
15:45:13 [<MYHOST> 3][ERROR]                           separate git dir from working tree
15:45:13 [<MYHOST> 3][ERROR]     -c, --config <key=value>
15:45:13 [<MYHOST> 3][ERROR]                           set config inside the new repository
15:45:13 [<MYHOST> 3][ERROR] 
15:45:13 [<MYHOST> 3][ERROR] Exit code: 129
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/util/subprocess.rb:75:in `execute'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/shellgit/base_repository.rb:98:in `git'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/shellgit/bare_repository.rb:19:in `clone'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/cache.rb:91:in `sync!'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/cache.rb:71:in `sync'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/stateful_repository.rb:33:in `sync'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:48:in `visit_module'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:34:in `visit'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/module/base.rb:80:in `accept'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:94:in `block (2 levels) in accept'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:93:in `each'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:93:in `block in accept'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:42:in `visit_puppetfile'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:34:in `visit'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:92:in `accept'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:29:in `call'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/runner.rb:30:in `call'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/cri_runner.rb:67:in `call'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command_dsl.rb:223:in `block in runner'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:298:in `call'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:298:in `run_this'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:251:in `run'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:264:in `run'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:264:in `run'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/bin/r10k:7:in `<top (required)>'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/bin/r10k:23:in `load'
15:45:13 [<MYHOST> 3][ERROR] C:/ProgramData/ruby/bin/r10k:23:in `<main>'
15:45:13 [<MYHOST> 3][ERROR] ERROR   -> Command exited with non-zero exit code:
15:45:13 [<MYHOST> 3][ERROR] Command: git clone --mirror https://github.com/badgerious/puppet-windows-env.git C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-badgerious-puppet-windows-env.git
15:45:13 [<MYHOST> 3][ERROR] Stderr:
15:45:13 [<MYHOST> 3][ERROR] Too many arguments.
15:45:13 [<MYHOST> 3][ERROR] 
15:45:13 [<MYHOST> 3][ERROR] usage: git clone [options] [--] <repo> [<dir>]
```

**Comment by adrienthebo** [comment]
What version of Git and r10k are you using? Could you add the `-v debug2 flag` to the command and provide the output of that as well?

**Comment by adrienthebo** [comment]
And if you run r10k on the host itself it works, while running it over ssh fails?

**Comment by lmayorga1980** [comment]
Yep. 

Successful run from the windows host

https://gist.github.com/lmayorga1980/8bb9daf3c7aede5fc4f5

But it fails when executing from SSH connection.

**Comment by lmayorga1980** [comment]
Wondering if the SSH connection breaks certain Environment Variables

https://gist.github.com/lmayorga1980/2420b331ebd2326a10c3

**Comment by adrienthebo** [comment]
Do you have multiple versions of Git installed? The difference between having an interactive session and just running a command could give you a different path.

**Comment by lmayorga1980** [comment]
Yeah I guess I am going to try it with the full PATH. Wondering if there is a way for r10k to understand something like **R10K_GITPATH**

![screen shot 2015-06-17 at 12 52 52 pm](https://cloud.githubusercontent.com/assets/3172378/8213463/f9faf250-14ef-11e5-9321-9d411044786f.png)

**Comment by adrienthebo** [comment]
It looks like the same Git binary is being used for both the interactive r10k case and the remote r10k case, so it doesn't look like that's the issue. The error reporting should print the git command that's failing, could you try running that command over SSH to see if the isolated Git command fails?

**Comment by lmayorga1980** [comment]
https://gist.github.com/lmayorga1980/04ddfd31ad7e042f962a

**Comment by adrienthebo** [comment]
I wouldn't expect running r10k over SSH to cause this, but since Windows uses `Open3.capture3` to run commands on Windows there might be issue with quoting fields. Could you try moving the module install directory into `C:\r10k-temp` or something like that to avoid issues with the space in `"Program Files"` ?

**Comment by lmayorga1980** [comment]
Just removed the custom `module_dir`

```
mod 'puppetlabs-stdlib',
  :git => 'https://github.com/puppetlabs/puppetlabs-stdlib.git',
  :tag => '4.6.0'

mod 'badgerious-windows_env',
  :git => 'https://github.com/badgerious/puppet-windows-env.git',
  :tag => 'v2.2.1'

mod 'puppetlabs-registry',
  :git => 'https://github.com/puppetlabs/puppetlabs-registry.git',
  :tag => '1.1.0'

mod 'puppetlabs-vcsrepo',
  :git => 'https://github.com/puppetlabs/puppetlabs-vcsrepo.git',
  :tag => '1.3.0'

mod 'chocolatey-chocolatey',
  :git => 'https://github.com/chocolatey/puppet-chocolatey.git',
  :tag => '1.0.0'
```

But still fails.

```
[2015-06-18 12:26:11 - DEBUG1] Testing to see if feature shellgit is available.
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Evaluating proc #<Proc:0x1248648@C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/features.rb:16> to test for feature shellgit
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Proc #<Proc:0x1248648@C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/features.rb:16> for feature shellgit returned "C:\\Program Files (x86)\\Git\\bin/git.EXE"
12:26:11            [2015-06-18 12:26:11 - DEBUG1] Feature shellgit is available.
12:26:11            [2015-06-18 12:26:11 - DEBUG1] Testing to see if feature shellgit is available.
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Evaluating proc #<Proc:0x1248648@C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/features.rb:16> to test for feature shellgit
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Proc #<Proc:0x1248648@C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/features.rb:16> for feature shellgit returned "C:\\Program Files (x86)\\Git\\bin/git.EXE"
12:26:11            [2015-06-18 12:26:11 - DEBUG1] Feature shellgit is available.
12:26:11            [2015-06-18 12:26:11 - INFO] Updating module C:/puppet/modules/stdlib
12:26:11            [2015-06-18 12:26:11 - DEBUG1] Creating new git cache for "https://github.com/puppetlabs/puppetlabs-stdlib.git"
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Starting process: ["git", "clone", "--mirror", "https://github.com/puppetlabs/puppetlabs-stdlib.git", "C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-puppetlabs-puppetlabs-stdlib.git"]
12:26:11            [2015-06-18 12:26:11 - DEBUG2] Finished process:
12:26:11            Command: git clone --mirror https://github.com/puppetlabs/puppetlabs-stdlib.git C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-puppetlabs-puppetlabs-stdlib.git
12:26:11            Stderr:
12:26:11            Too many arguments.
```

Wondering if there is a way to switch the default mirror directory

**Comment by lmayorga1980** [comment]
https://github.com/puppetlabs/r10k/blob/0e1dae13a6fe5db39f8a6e1ed941039fec818cc4/lib/r10k/git/cache.rb#L19

![screen shot 2015-06-18 at 12 50 31 pm](https://cloud.githubusercontent.com/assets/3172378/8236796/ac7ddf5a-15b8-11e5-88ae-d93628fcf21b.png)

Wondering if i can switch that VAR temporarily for the sake of r10k.

**Comment by lmayorga1980** [comment]
Take my last comment out. I believe when expanded the `Program files` on the **Cache** class is the issue. 

Might be wrong of course.

**Comment by lmayorga1980** [comment]
Ok I was able to change add the following lines in order to clone the repository.

on 

https://github.com/puppetlabs/r10k/blob/0e1dae13a6fe5db39f8a6e1ed941039fec818cc4/lib/r10k/git/shellgit/bare_repository.rb

https://github.com/puppetlabs/r10k/blob/0e1dae13a6fe5db39f8a6e1ed941039fec818cc4/lib/r10k/git/rugged/bare_repository.rb#L18-L20

``` ruby
  def git_dir
    @path
    @path = "\"#{@path}\""
  end
```

Now I get a different error after cloning ...

```
15-06-18 16:15:25 - INFO] Updating module C:/puppet/modules/chocolatey
16:15:25            [2015-06-18 16:15:25 - DEBUG1] Creating new git cache for "https://github.com/chocolatey/puppet-chocolatey.git"
16:15:25            [2015-06-18 16:15:25 - DEBUG2] Starting process: ["git", "clone", "--mirror", "https://github.com/chocolatey/puppet-chocolatey.git", "\"C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git\""]
16:15:26            [2015-06-18 16:15:25 - DEBUG2] Finished process:
16:15:26            Command: git clone --mirror https://github.com/chocolatey/puppet-chocolatey.git "C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git"
16:15:26            Stderr:
16:15:26            Cloning into bare repository 'C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git'...
16:15:26            Exit code: 0
16:15:26            [2015-06-18 16:15:25 - DEBUG2] Starting process: ["git", "--git-dir", "\"\"C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git\"\"", "rev-parse", "1.0.0^{commit}"]
16:15:26            [2015-06-18 16:15:25 - DEBUG2] Finished process:
16:15:26            Command: git --git-dir ""C:/Program Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git"" rev-parse 1.0.0^{commit}
16:15:26            Stderr:
16:15:26            git: 'Files/OpenSSH/home/_puppet/.r10k/git/https---github.com-chocolatey-puppet-chocolatey.git' is not a git command. See 'git --help'.
16:15:26            Exit code: 1
16:15:26            [2015-06-18 16:15:25 - ERROR] Unable to sync repo to unresolvable ref '1.0.0' at C:/puppet/modules/chocolatey/.git
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/git/stateful_repository.rb:38:in `sync'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:48:in `visit_module'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:34:in `visit'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/module/base.rb:80:in `accept'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:94:in `block (2 levels) in accept'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:93:in `each'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:93:in `block in accept'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:42:in `visit_puppetfile'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:34:in `visit'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/puppetfile.rb:92:in `accept'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/puppetfile/install.rb:29:in `call'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/runner.rb:30:in `call'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/lib/r10k/action/cri_runner.rb:67:in `call'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command_dsl.rb:223:in `block in runner'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:298:in `call'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:298:in `run_this'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:251:in `run'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:264:in `run'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/cri-2.6.1/lib/cri/command.rb:264:in `run'
16:15:26            C:/ProgramData/ruby/lib/ruby/gems/2.1.0/gems/r10k-2.0.1/bin/r10k:7:in `<top (required)>'
16:15:26            C:/ProgramData/ruby/bin/r10k:23:in `load'
16:15:26            C:/ProgramData/ruby/bin/r10k:23:in `<main>'
```

**Comment by adrienthebo** [comment]
The patch that you've added will add quotes each time that method is called, which isn't what you want. Try applying this patch:

``` diff
diff --git i/lib/r10k/util/subprocess/runner/windows.rb w/lib/r10k/util/subprocess/runner/windows.rb
index 7ddcd42..a5f0b69 100644
--- i/lib/r10k/util/subprocess/runner/windows.rb
+++ w/lib/r10k/util/subprocess/runner/windows.rb
@@ -14,7 +14,13 @@ class R10K::Util::Subprocess::Runner::Windows < R10K::Util::Subprocess::Runner
   end

   def run
-    cmd = @argv.join(' ')
+    cmd = @argv.map do |elem|
+      if elem.match(/\s/)
+        %Q["#{elem}"]
+      else
+        elem
+      end
+    end.join(' ')
```

**Comment by lmayorga1980** [comment]
Are you planning to include this change on the following release?

**Comment by adrienthebo** [comment]
Your fix or the patch that I provided above?

**Comment by lmayorga1980** [comment]
Your patch ;)

**Comment by adrienthebo** [comment]
If it works then I'll include it, but it's a fairly big task for me to do testing on Windows so external validation of that patch would help a lot.

**Comment by lmayorga1980** [comment]
Then I will validated on Windows. Are u creating a branch?

**Comment by lmayorga1980** [comment]
The fix works but not sure to target it to the master branch https://github.com/puppetlabs/r10k/pull/458

**Comment by adrienthebo** [comment]
Closed by #469.

**Comment by adrienthebo** [comment]
See also https://tickets.puppetlabs.com/browse/RK-143.

**Comment by lmayorga1980** [comment]
broken again in 2.0.3

**Comment by adrienthebo** [comment]
The fix for this was merged into the master branch, and thus has not been released - it'll be available in 2.1.0 which should be out in a week or two.
